File: leetcode/AIcoding/1.Gemini-2.5/c_solutions/1105-filling-bookcase-shelves.c
#include <limits.h>

int max(int a, int b) {
    return a > b ? a : b;
}

int min(int a, int b) {
    return a < b ? a : b;
}

int minHeightShelves(int** books, int booksSize, int* booksColSize, int shelfWidth) {
    int n = booksSize;
    int dp[n + 1];

    dp[0] = 0;

    for (int i = 1; i <= n; i++) {
        dp[i] = INT_MAX;
    }

    for (int i = 1; i <= n; i++) {
        int current_width = 0;
        int max_height_on_shelf = 0;

        for (int j = i - 1; j >= 0; j--) {
            current_width += books[j][0];
            max_height_on_shelf = max(max_height_on_shelf, books[j][1]);

            if (current_width <= shelfWidth) {
                dp[i] = min(dp[i], dp[j] + max_height_on_shelf);
            } else {
                break;
            }
        }
    }

    return dp[n];
}

File: leetcode/AIcoding/5.Qwen-3/go_solutions/1105-filling-bookcase-shelves.go
package main

func minHeightShelves(books [][]int, shelfWidth int) int {
    n := len(books)
    dp := make([]int, n+1)
    for i := 1; i <= n; i++ {
        maxH := 0
        dp[i] = 1 << 31
        for j := i - 1; j >= 0 && shelfWidth >= books[j][0]; j-- {
            maxH = max(maxH, books[j][1])
            dp[i] = min(dp[i], dp[j]+maxH)
        }
    }
    return dp[n]
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}